### PR TITLE
20191106 postgres dbapi

### PR DESCRIPTION
**Pull request by peterthomassen**

### CODE REVIEW

**Comment by peterthomassen** [comment]
Migration instructions (tested on sandbox):

1. `docker-compose stop api celery`
2. `docker-compose run --rm api python manage.py dumpdata contenttypes --indent=4 --natural-foreign > contenttype.json`
3. `docker-compose run --rm api python manage.py dumpdata --exclude contenttypes --exclude desecapi.Donation --exclude desecapi.AuthenticatedAction --exclude desecapi.AuthenticatedBasicUserAction --exclude desecapi.AuthenticatedUserAction --exclude desecapi.AuthenticatedActivateUserAction --exclude desecapi.AuthenticatedChangeEmailUserAction --exclude desecapi.AuthenticatedResetPasswordUserAction --exclude desecapi.AuthenticatedDeleteUserAction --exclude desecapi.AuthenticatedDomainBasicUserAction --exclude desecapi.AuthenticatedRenewDomainBasicUserAction --indent=4 --natural-foreign > everything_else.json`
4. `docker-compose stop dbapi`

5. `git pull && docker-compose pull  # pre-merge: git checkout 20191106_postgres_dbapi && docker-compose build api dbapi`

6. `docker-compose up -d dbapi`
7. `docker-compose run --rm api bash`
8. in another terminal, run:
    1. `docker ps  # find container name like desec-stack_api_run_*`
    2. `docker cp contenttype.json desec-stack_api_run_XXX:/usr/src/app/`
    3. `docker cp everything_else.json desec-stack_api_run_XXX:/usr/src/app/`
9. back in the `api` shell, run:
    ```
    python manage.py migrate
    python manage.py sqlflush | ./manage.py dbshell
    python manage.py loaddata contenttype.json
    python manage.py loaddata everything_else.json
    rm contenttype.json everything_else.json
    exit
    ```
10. `docker-compose up -d`
11. If all went well: `rm contenttype.json everything_else.json`

File: README.md [hunk 1]
```diff
@@ -135,10 +135,10 @@ While there are certainly many ways to get started hacking desec-stack, here is
     For desec-stack, [docker](https://docs.docker.com/install/linux/docker-ce/ubuntu/) and [docker-compose](https://docs.docker.com/compose/install/) are required.
     Further tools that are required to start hacking are git and curl.
     Recommended, but not strictly required for desec-stack development is to use certbot along with Let's Encrypt and PyCharm.
-    jq, httpie, libmariadbclient-dev, python3-dev (>= 3.8) and python3-venv (>= 3.8) are useful if you want to follow this guide.
+    jq, httpie, libpq-dev, python3-dev (>= 3.8) and python3-venv (>= 3.8) are useful if you want to follow this guide.
```
Comment (nils-wisiol) [inline README.md:138]: install both clients, mariadb and pq

File: README.md [hunk 2]
```diff
@@ -135,10 +135,10 @@ While there are certainly many ways to get started hacking desec-stack, here is
...
     The webapp requires nodejs. To install everything you need for this guide except docker and docker-compose, use
 
-       sudo apt install certbot curl git httpie jq libmariadbclient-dev nodejs npm python3-dev python3-venv libmemcached-dev
+       sudo apt install certbot curl git httpie jq libpq-dev nodejs npm python3-dev python3-venv libmemcached-dev
```
Comment (nils-wisiol) [inline README.md:141]: install both clients, mariadb and pq

File: api/Dockerfile [hunk 3]
```diff
@@ -9,7 +9,7 @@ ENV PIP_DISABLE_PIP_VERSION_CHECK=1
 ENV PIP_NO_CACHE_DIR=1
 
 COPY requirements.txt /usr/src/app/
-RUN apk add --no-cache gcc freetype-dev libffi-dev musl-dev libmemcached-dev mariadb-connector-c-dev jpeg-dev zlib-dev \
+RUN apk add --no-cache gcc freetype-dev libffi-dev musl-dev libmemcached-dev postgresql-dev jpeg-dev zlib-dev \
```
Comment (nils-wisiol) [inline api/Dockerfile:12]: why this Dockerfile has two apk add commands?

File: api/api/settings.py [hunk 4]
```diff
@@ -62,21 +62,12 @@
...
         'PASSWORD': os.environ['DESECSTACK_DBAPI_PASSWORD_desec'],
         'HOST': 'dbapi',
-        'OPTIONS': {
-            'charset': 'utf8mb4',
-            'init_command': "SET sql_mode='STRICT_TRANS_TABLES'",
-        },
-        'TEST': {
-            'CHARSET': 'utf8mb4',
-            'COLLATION': 'utf8mb4_bin',
-        },
```
Comment (nils-wisiol) [inline api/api/settings.py:77]: let's make a test to see if the migration works correctly for utf8 content

File: api/desecapi/serializers.py [hunk 5]
```diff
@@ -494,12 +494,6 @@ def is_empty(data_item):
         # so for parallel requests, we can get integrity errors due to duplicate keys.
         # This will be considered a 429-error, even though re-sending the request will be successful.
```
Comment (nils-wisiol) [inline api/desecapi/serializers.py:495]: adjust comment according to change below

File: dbapi/pg_hba.conf [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+#host    all             all             127.0.0.1/32            scram-sha-256
+# IPv6 local connections:
+#host    all             all             ::1/128                 scram-sha-256
+# Allow replication connections from localhost, by a user with the
+# replication privilege.
+#local   replication     all                                     trust
+#host    replication     all             127.0.0.1/32            scram-sha-256
+#host    replication     all             ::1/128                 scram-sha-256
+
+host desec desec all scram-sha-256
```
Comment (nils-wisiol) [inline dbapi/pg_hba.conf:15]: let's test the permissions on sandbox

**Review by nils-wisiol — APPROVED** [review]
lgtm, but some minor changes are requested

File: dbapi/pg_hba.conf (hunk 6 above)
Comment (peterthomassen) [reply dbapi/pg_hba.conf:15]: done (locally)

File: README.md (hunk 1 above)
Comment (peterthomassen) [reply README.md:138]: done

File: README.md (hunk 2 above)
Comment (peterthomassen) [reply README.md:141]: done

File: api/desecapi/serializers.py (hunk 5 above)
Comment (peterthomassen) [reply api/desecapi/serializers.py:495]: done

File: api/api/settings.py (hunk 4 above)
Comment (peterthomassen) [reply api/api/settings.py:77]: checked

**Pull request merged**